File: software/demo_ghostprovider/analyzer/probe.py
"""System probing — prerequisite checks, network scan, port fingerprinting."""

import re
import shutil
import socket
import subprocess
from typing import Any

from demo_ghostprovider.analyzer.models import (
    AnalysisResult,
    InterfaceInfo,
    ListeningPort,
    NetworkInfo,
    ServiceFingerprint,
)
from demo_ghostprovider.analyzer.signatures import SERVICE_SIGNATURES


def fingerprint_port(port: int, proto: str = "tcp") -> ServiceFingerprint | None:
    """Try to fingerprint an HTTP service on a given port."""
    try:
        with socket.create_connection(("127.0.0.1", port), timeout=3) as sock:
            sock.settimeout(5)
            sock.sendall(b"GET / HTTP/1.0\r\nHost: localhost\r\n\r\n")
            response = sock.recv(8192)
    except (TimeoutError, OSError):
        return None

    headers_end = response.find(b"\r\n\r\n")
    if headers_end == -1:
        return None

    body = response[headers_end + 4:]
    headers_raw = response[:headers_end].decode("utf-8", errors="replace")
    status_line = headers_raw.split("\r\n")[0] if headers_raw else ""

    details: dict[str, Any] = {
        "status_line": status_line,
        "server_header": "",
    }

    for line in headers_raw.split("\r\n")[1:]:
        if line.lower().startswith("server:"):
            details["server_header"] = line.split(":", 1)[1].strip()
            break

    # Match against known signatures
    for sig_pattern, svc_type, svc_name, confidence in SERVICE_SIGNATURES:
        if sig_pattern.search(body):
            return ServiceFingerprint(
                port=port,
                proto=proto,
                service_type=svc_type,
                service_name=svc_name,
                confidence=confidence,
                details=details,
            )

    # Fallback: detect generic web server from server header
    server = details.get("server_header", "").lower()
    if server:
        if any(x in server for x in ("nginx", "apache", "caddy", "iis")):
            return ServiceFingerprint(
                port=port, proto=proto,
                service_type="web_app",
                service_name=server.split("/")[0].title(),
                confidence=60,
                details=details,
            )
        if "gunicorn" in server:
            return ServiceFingerprint(
                port=port, proto=proto,
                service_type="web_app",
                service_name="Python WSGI (gunicorn)",
                confidence=70,
                details=details,
            )
        if "uvicorn" in server:
            return ServiceFingerprint(
                port=port, proto=proto,
                service_type="web_app",
                service_name="Python ASGI (uvicorn)",
                confidence=70,
                details=details,
            )
        if "node" in server.lower() or "express" in server.lower():
            return ServiceFingerprint(
                port=port, proto=proto,
                service_type="web_app",
                service_name="Node.js HTTP Server",
                confidence=65,
                details=details,
            )
        if "python" in server.lower():
            return ServiceFingerprint(
                port=port, proto=proto,
                service_type="web_app",
                service_name="Python HTTP Server",
                confidence=60,
                details=details,
            )

    # Generic HTTP response → unknown web app
    return ServiceFingerprint(
        port=port, proto=proto,
        service_type="web_app",
        service_name="Unknown HTTP Service",
        confidence=30,
        details=details,
    )


def _fingerprint_all_services(ports: list[ListeningPort]) -> list[ServiceFingerprint]:
    services: list[ServiceFingerprint] = []
    for p in ports:
        fp = fingerprint_port(p.port, p.proto)
        if fp is not None:
            services.append(fp)
    return services


def _check_cmd(name: str) -> bool:
    return shutil.which(name) is not None


def _check_localhost() -> bool:
    try:
        with socket.create_connection(("127.0.0.1", 80), timeout=2):
            return True
    except (OSError, ConnectionRefusedError):
        pass
    try:
        with socket.create_connection(("127.0.0.1", 8080), timeout=2):
            return True
    except (OSError, ConnectionRefusedError):
        return False


def _check_systemd_nspawn() -> bool:
    """Check if systemd-nspawn is available."""
    try:
        r = subprocess.run(
            ["systemd-nspawn", "--version"],
            capture_output=True, timeout=5,
            check=False,
        )
        return r.returncode == 0
    except (FileNotFoundError, subprocess.TimeoutExpired):
        return False


def _check_network() -> bool:
    try:
        subprocess.run(
            ["ping", "-c", "1", "-W", "2", "8.8.8.8"],
            capture_output=True, timeout=5,
            check=False,
        )
        return True
    except (subprocess.TimeoutExpired, FileNotFoundError):
        return False


def _detect_interfaces() -> list[InterfaceInfo]:
    interfaces: list[InterfaceInfo] = []
    try:
        result = subprocess.run(
            ["ip", "-br", "addr", "show"],
            capture_output=True, text=True, timeout=5,
            check=False,
        )
        if result.returncode == 0:
            for line in result.stdout.strip().split("\n"):
                parts = line.split()
                if len(parts) >= 3:
                    name = parts[0]
                    status = "up" if parts[1] == "UP" else "down"
                    ip_info = parts[2] if len(parts) > 2 else ""
                    ip = ip_info.split("/")[0] if ip_info else ""
                    netmask = f"/{ip_info.split('/')[1]}" if "/" in ip_info else ""
                    interfaces.append(InterfaceInfo(
                        name=name, ip=ip, netmask=netmask, status=status,
                    )
                    )
    except (subprocess.TimeoutExpired, FileNotFoundError):
        pass
    return interfaces


def _detect_listening_ports() -> list[ListeningPort]:
    ports: list[ListeningPort] = []
    try:
        result = subprocess.run(
            ["ss", "-tlnp4"],
            capture_output=True, text=True, timeout=5,
            check=False,
        )
        if result.returncode == 0:
            for line in result.stdout.strip().split("\n")[1:]:
                if not line.strip():
                    continue
                parts = line.split()
                if len(parts) >= 4:
                    proto = "tcp"
                    addr_port = parts[3]
                    process = ""
                    if len(parts) > 4:
                        proc_match = re.search(r'users:\(\("(.+?)"', parts[-1])
                        if proc_match:
                            process = proc_match.group(1)
                    if ":" in addr_port:
                        addr, port_str = addr_port.rsplit(":", 1)
                        try:
                            port = int(port_str)
                            ports.append(ListeningPort(
                                port=port, proto=proto,
                                address=addr, process=process,
                            )
                            )
                        except ValueError:
                            pass
    except (subprocess.TimeoutExpired, FileNotFoundError):
        pass
    return ports


def _detect_vpn(interfaces: list[InterfaceInfo]) -> tuple[bool, list[str]]:
    vpn_keywords = {"tun", "tap", "wg", "ppp", "vpn", "virbr"}
    vpn_ifaces = []
    for iface in interfaces:
        name_lower = iface.name.lower()
        if any(kw in name_lower for kw in vpn_keywords):
            vpn_ifaces.append(iface.name)
    return (len(vpn_ifaces) > 0), vpn_ifaces


def _get_gateway() -> str:
    try:
        result = subprocess.run(
            ["ip", "route", "show", "default"],
            capture_output=True, text=True, timeout=5,
            check=False,
        )
        if result.returncode == 0 and result.stdout.strip():
            parts = result.stdout.strip().split()
            if len(parts) >= 3:
                return parts[2]
    except (subprocess.TimeoutExpired, FileNotFoundError):
        pass
    return ""


def _get_dns() -> list[str]:
    dns_servers = []
    try:
        with open("/etc/resolv.conf") as f:
            for line in f:
                if line.startswith("nameserver "):
                    dns_servers.append(line.split()[1])
    except (FileNotFoundError, OSError):
        pass
    return dns_servers


def run_analysis() -> AnalysisResult:
    interfaces = _detect_interfaces()
    ports = _detect_listening_ports()
    vpn_active, vpn_ifaces = _detect_vpn(interfaces)

    services = _fingerprint_all_services(ports)

    net_info = NetworkInfo(
        interfaces=interfaces,
        listening_ports=ports,
        services=services,
        vpn_active=vpn_active,
        vpn_interfaces=vpn_ifaces,
        gateway=_get_gateway(),
        dns=_get_dns(),
    )

    result = AnalysisResult(
        systemd=_check_cmd("systemctl"),
        systemd_nspawn=_check_systemd_nspawn(),
        git=_check_cmd("git"),
        python3=_check_cmd("python3"),
        node=_check_cmd("node"),
        localhost=_check_localhost(),
        network=_check_network(),
        network_info=net_info,
    )
    if not result.systemd:
        result.errors.append("systemd not found — required for service management")
    if not result.systemd_nspawn:
        result.errors.append("systemd-nspawn not found — recommended for isolated hosting")
    if not result.git:
        result.errors.append("Git not found — cannot clone repositories")
    if not result.network:
        result.errors.append("No network — cannot fetch remote repositories")
    return result

File: src/analyzer/probe.rs
//! System probing: prerequisite checks and a bare map of occupied ports.
//!
//! Privacy policy: this report must stay useless to attackers. No VPN
//! interface detection, no HTTP probing of local services — the port table
//! names only what `ss` itself already shows to the local user.

use std::net::{Ipv4Addr, SocketAddr, SocketAddrV4, TcpStream};
use std::process::Command;
use std::time::Duration;

use super::models::{AnalysisResult, InterfaceInfo, ListeningPort};

fn which(bin: &str) -> bool {
    std::env::var_os("PATH")
        .map(|p| std::env::split_paths(&p).any(|dir| dir.join(bin).is_file()))
        .unwrap_or(false)
}

fn detect_interfaces() -> Vec<InterfaceInfo> {
    let Ok(out) = Command::new("ip").args(["-br", "addr", "show"]).output() else {
        return vec![];
    };
    if !out.status.success() {
        return vec![];
    }
    String::from_utf8_lossy(&out.stdout)
        .lines()
        .filter_map(|line| {
            let parts: Vec<&str> = line.split_whitespace().collect();
            if parts.len() < 2 {
                return None;
            }
            let name = parts[0].to_string();
            let status = if parts[1] == "UP" || parts[1] == "UNKNOWN" {
                "up"
            } else {
                "down"
            }
            .to_string();
            let ip_info = parts.get(2).copied().unwrap_or("");
            let ip = ip_info.split('/').next().unwrap_or("").to_string();
            let netmask = format!("/{}", ip_info.split('/').nth(1).unwrap_or(""));
            Some(InterfaceInfo {
                name,
                ip,
                netmask,
                status,
            })
        })
        .collect()
}

/// Allowlisted, net.log-recorded network probe: an HTTPS GET to github.com
/// (the host every fetch in this tool actually needs). Deliberately NOT
/// ICMP ping or raw DNS resolution — those would be outbound contacts the
/// transparency guarantees do not cover.
fn https_probe_ok() -> bool {
    crate::hoster::httpclient::get_text("https://github.com/").is_ok()
}

/// Parse one `ss -tlnp` data row.
///
/// Owner attribution never happens here — a bare port/address pair only.
/// Which service owns a port is resolved exclusively from local state.json
/// (deployments made via this panel), never from `ss`.
pub fn parse_ss_row(line: &str) -> Option<ListeningPort> {
    let parts: Vec<&str> = line.split_whitespace().collect();
    if parts.len() < 4 {
        return None;
    }
    let addr_port = parts[3];
    let (_, port_str) = addr_port.rsplit_once(':')?;
    let port = port_str.parse().ok()?;
    Some(ListeningPort {
        port,
        address: addr_port.to_string(),
    })
}

/// All TCP listeners, both address families.
///
/// Deployed services frequently bind an IPv6 wildcard (`*:port`, accepting
/// v4-mapped connections) — filtering with `-4` silently hid exactly the
/// ports this panel had assigned. No family filter, ever.
fn detect_listening_ports() -> Vec<ListeningPort> {
    let Ok(out) = Command::new("ss").args(["-tlnp"]).output() else {
        return vec![];
    };
    if !out.status.success() {
        return vec![];
    }
    String::from_utf8_lossy(&out.stdout)
        .lines()
        .skip(1)
        .filter_map(parse_ss_row)
        .collect()
}

fn get_gateway() -> String {
    let Ok(out) = Command::new("ip")
        .args(["route", "show", "default"])
        .output()
    else {
        return String::new();
    };
    String::from_utf8_lossy(&out.stdout)
        .split_whitespace()
        .nth(2)
        .unwrap_or("")
        .to_string()
}

fn get_dns() -> Vec<String> {
    let Ok(content) = std::fs::read_to_string("/etc/resolv.conf") else {
        return vec![];
    };
    content
        .lines()
        .filter_map(|l| l.strip_prefix("nameserver ").map(|s| s.trim().to_string()))
        .collect()
}

/// Run the full local analysis.
pub fn run_analysis() -> AnalysisResult {
    let interfaces = detect_interfaces();
    let ports = detect_listening_ports();

    let localhost_ok = [80u16, 8080].iter().any(|p| {
        TcpStream::connect_timeout(
            &SocketAddr::from(SocketAddrV4::new(Ipv4Addr::LOCALHOST, *p)),
            Duration::from_secs(2),
        )
        .is_ok()
    });

    let mut result = AnalysisResult {
        systemd: which("systemctl"),
        systemd_nspawn: which("systemd-nspawn"),
        git: which("git"),
        python3: which("python3"),
        node: which("node"),
        localhost: localhost_ok,
        // Everything this tool fetches lives behind github.com, so a
        // reachable github.com is the honest definition of "network".
        network: https_probe_ok(),
        interfaces,
        listening_ports: ports,
        gateway: get_gateway(),
        dns: get_dns(),
        errors: vec![],
    };

    if !result.systemd {
        result
            .errors
            .push("systemd not found — required for service management".into());
    }
    if !result.systemd_nspawn {
        result
            .errors
            .push("systemd-nspawn not found — recommended for isolated hosting".into());
    }
    if !result.git {
        result
            .errors
            .push("Git not found — cannot clone repositories".into());
    }
    if !result.network {
        result
            .errors
            .push("No network — cannot fetch remote repositories".into());
    }
    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ss_row_parses_bare_port_and_address() {
        let line = "LISTEN 0      128        0.0.0.0:5222       0.0.0.0:*    users:((\"socat\",pid=1234,fd=5))";
        let p = parse_ss_row(line).expect("parses");
        assert_eq!(p.port, 5222);
        assert_eq!(p.address, "0.0.0.0:5222");
    }

    /// Deployed services usually bind an IPv6 wildcard; such rows are the
    /// whole point of the port table and must parse.
    #[test]
    fn ss_row_ipv6_wildcard_is_parsed() {
        let line = "LISTEN 0      4096       *:23920             *:*    users:((\"ghost-server\",pid=692739,fd=9))";
        let p = parse_ss_row(line).expect("parses");
        assert_eq!(p.port, 23920);
        assert_eq!(p.address, "*:23920");

        let bracketed = "LISTEN 0 4096 [::]:8080 [::]:* users:((\"node\",pid=1,fd=7))";
        let p = parse_ss_row(bracketed).expect("parses");
        assert_eq!(p.port, 8080);
    }

    /// Sockets owned by other users parse the same way — no owner data is
    /// ever recorded for foreign listeners.
    #[test]
    fn ss_row_without_owner_still_yields_port() {
        let line = "LISTEN 0      200        127.0.0.1:5432       0.0.0.0:*";
        let p = parse_ss_row(line).expect("parses");
        assert_eq!(p.port, 5432);
        assert_eq!(p.address, "127.0.0.1:5432");
    }

    #[test]
    fn ss_row_garbage_is_rejected() {
        assert!(parse_ss_row("").is_none());
        assert!(parse_ss_row("State Recv-Q Send-Q").is_none());
        assert!(parse_ss_row("LISTEN 0 0 notaport 0.0.0.0:*").is_none());
    }
}
